File: Dongwon/Test_question/PriviousTest_Kakao/2019/hotel.js
const solution = (x, room_number) => {
  let rooms = new Map(),
    answer = [],
    tt = [];
  for (let i = 0; i < room_number.length; i++) {
    const get_num = rooms.has(room_number[i]);
    if (get_num) {
      let temp = [room_number[i]];
      let number = room_number[i];

      while (true) {
        let idx = number;
        number = rooms.get(idx);
        if (!number) {
          rooms.set(idx, idx + 1);
          answer.push(idx);
          for (let i = 0; i < temp.length; i++) {
            rooms.set(temp[i], idx + 1);
          }
          break;
        }
        temp.push(number);
      }
    } else {
      answer.push(room_number[i]);
      rooms.set(room_number[i], room_number[i] + 1);
    }
  }
  return answer;
};

File: Dongwon/Test_question/PriviousTest_Kakao/2019/hotel.py
# 1차 실패 풀이

def solution(k, room_number):
    result = []
    used = [False for __ in range(k+1)]
    room_dict = dict()
    for i in range(1, k+1):
        room_dict[i] = i
    
    for item in room_number:
        # print(room_dict, item)
        # print(used,used[item], result)
        if not used[item]:
            used[item] = True
            result.append(item)
            temp = item
            while used[temp]:
                temp += 1
            room_dict[item] = temp
        else:
            temp = item
            while used[temp]:
                temp += 1
            
            result.append(room_dict[temp])
            used[room_dict[temp]] = True
            room_dict[temp] = temp
        
    return result

# 2차 풀이

def solution2(k, room_number):
    result = []
    room_dict = dict()
    
    for number in room_number:
        num = room_dict.get(number, 0)
        if num:
            temp = [number]
            while True:
                idx = num
                num = room_dict.get(num, 0)
                if not num:
                    result.append(idx)
                    room_dict[idx] = idx + 1
                    for i in temp:
                        room_dict[i] = idx + 1
                    break
                temp.append(num)
        else:
            result.append(number)
            room_dict[number] = number + 1
        
    return result